### ISSUE TITLE
proposal: Limit support to one GOOS value (primarily for stdlib).

### ISSUE THREAD

**Issue by dmitshur** [gopherjs/gopherjs#693]
The GopherJS compiler supports building output only for GOARCH=js value, and any GOOS value.

That means it will produce valid and different output for this Go package, depending on which GOOS value is used:

```Go
// +build plan9

package p

const Message = "It was a hot summer..."
```

```Go
// +build !plan9

package p

const Message = "It was a cold winter..."
```

Of course, it also supports cross-compiling, so one can produce different outputs by doing:

```bash
$ GOOS=plan9 gopherjs build
# hot summer

$ GOOS=solaris gopherjs build
# cold winter
```

I propose we limit support to just a single fixed GOOS value.

This is primarily meant to apply to the Go standard library, which we must manually apply overrides for in order to get it to build and work correctly for GOARCH=js. See https://github.com/gopherjs/gopherjs/commits/master/compiler/natives/src. However, I think we should apply for all packages for simplicity/consistency. Applying the limit to just Go stdlib but not 3rd party packages would be weird and unnecessary.

The motivation is simple. It costs a lot of developer time to support Go standard library for GopherJS. Right now, I don't think this project can afford to continue to support all GOOS values. It's an open source project, with no funding, and most of the progress is made by volunteers contributing in their spare time in after work hours.

The value gained from supporting multiple GOOS values is small. The cost is prohibitively high.

The compiler will continue to be able to (theoretically) compile for many GOOS values, we're just not going to support GOOS values other than a single one that we pick.

**Comment by shurcooL** [comment]
This came up when I investigated what it'd take to fix #688.

I considered 3 alternative fixed GOOS values to use:

1. `darwin` - this is the same GOOS that the 2 primary GopherJS developers (@neelance and me) use when developing GopherJS and using it

2. `linux` - it's one of the most popular and generic `GOOS` values. Even `android` build tag piggybacks off it. Also, unlike `darwin`, it's open source. Given open source and proprietary, I'd rather we support an open source OS.

3. `nacl` - this is the GOOS environment that is most similar to `GOARCH=js`. In fact, there are a number of times where I base `GOARCH=js` implementations after NaCL ones. There are tests that could be simplified if we also targetted `nacl`.

Using `darwin` is the easiest and works best from my testing. It's what Richard and I have used all this time. I'm sure many other users also use it.

Using `linux` would require some work; it didn't work out of box for me.

Using `windows` (for reference) seems like it'd require even more work. It clearly doesn't work right now (see #688).

Using `nacl` would require some initial investment, but would pay off in simplicity in the long term, IMO.

Consider the following Go code in `syscall` package override to print to stdout:

https://github.com/gopherjs/gopherjs/blob/f7c56533adfac99e12ccdccfec04a466d2a778da/compiler/natives/src/syscall/syscall_unix.go#L57-L74

Here's the equivalent code if we targeted `nacl`:

```Go
func naclWrite(fd int, b []byte) int {
	switch fd {
	case 1, 2:
		printToConsole(b)
		return len(b)
	default:
		return -1
	}
}
```

See how much cleaner that is?

Of course, I haven't implemented everything for `nacl`, that was just a brief attempt, so there may be unexpected things. But it seems like the best option in the long run to me.

**Comment by shurcooL** [comment]
To conclude, my recommendation if this proposal is accepted is to target `darwin` in the short term. And to consider optionally targetting `nacl` instead in the long run.

@neelance What do you think?

**Comment by neelance** [comment]
The main reason for supporting multiple GOOS is so that `gopherjs test` can work on multiple platforms.

Could you please list some examples of support for multiple GOOS causing overhead? Is it mainly about windows?

**Comment by shurcooL** [comment]
> The main reason for supporting multiple GOOS is so that gopherjs test can work on multiple platforms.

I don't understand why it `gopherjs test` wouldn't work on multiple platforms if we target compilation for a single GOOS value. Can you explain?

**Comment by shurcooL** [comment]
Ok, I see now. It's about the [syscall module](https://github.com/gopherjs/gopherjs/blob/master/doc/syscalls.md), which is used to implement `os` for reading/writing files from on disk during `gopherjs test` time (i.e., when using `node` to execute the JavaScript in terminal).

Normally, JavaScript is portable and can be run on any OS with `node`, but in this case, we're using non-portable syscall module API for `os` implementation.

I'm going to put this proposal on hold and think more about this, given the new information (well, old, but remembered now).

**Comment by j7b** [comment]
In playing around with #688 it seemed life would be simpler the goroot surrogate was (nacl) libgo, which might be a good idea anyway for wasm-related reasons. It also occurs to me the syscall issue could be addressed by giving the node addon moderately more love and mandating it if you want to gohperjs test.

**Comment by nevkontakte** [comment]
I would like to revive discussion of this proposal. Now that the upstream support `js,wasm` architecture we could reduce the amount of standard library overlays we maintain by reusing some of that code.

To be coherent with the upstream I would propose using `GOOS=js GOARCH=ecmascript`, which reflects that we run in the same JS environment, but our generated code is, in fact, ES5. 

As a proof of concept for code reuse, I hacked gopherjs build system to use `js/wasm` for certain low-level packages (with the focus on `syscall` in particular) and it seems to compile just fine, which is great news. If I manage to get it pass tests, it would allow us to dispose of our custom node-syscall extension 🙂 The only problem is, using different build tags for different packages is quite ugly and leads to inconsistencies. I would really prefer to be able to add js,ecmascript to the build tags in the original sources. @dmitshur do you think this is something the Go Team would be open to?.. If not, we'd have to fork code just to change build tags, which feels kind of silly.

**Comment by dmitshur** [comment]
In the Go project, an area we wanted to explore is improving experience for out-of-tree ports to reduce the need for having all ports be added in-tree. GopherJS is a compiler for an out-of-tree port. If there’s a small and reasonable change in the main repo that can help here significantly, I think it would be worth proposing and considering.

However, I don’t quite understand why a change would need to happen in the Go tree. You said:

> The only problem is, using different build tags for different packages is quite ugly and leads to inconsistencies.

Can you elaborate on this?

**Comment by nevkontakte** [comment]
Sorry, I realize now that my original comment had a few logical leaps that are likely confusing to anyone who can't read my mind 😅

Here's my chain of reasoning:

 - Originally GopherJS needed host's GOOS to interact with the OS (file system, etc.) and there wasn't one GOOS value that would be portable in the context of nodejs.
 - The price for the ability to do that was extra complexity of supporting several values GOOS and, for example, not supporting GOOS=windows even though nodejs is presumably a cross-platform runtime.
 - Wasm support in the upstream solved that problem and the code you build with the upstream Go under `GOOS=js GOARCH=wasm` will run exactly the same on nodejs on every OS it supports.
 - That means if GopherJS could take advantage of standard library code that is there for wasm, we could converge on our own `GOOS=js GOARCH=ecmascript` and simplify some things on our side.
 - But Go standard library uses very targeted build tags `js,wasm` to guard the code that would work just as well for GopherJS, so to take advantage of it we have three options:
   - Use `GOOS=js GOARCH=wasm` for our builds. That's probably not ideal for our users who may want to distinguish between GopherJS and wasm at build time.
   - Hard-code exceptions in our build process to adjust the set of sources we build despite the build tags.
   - Change standard library sources upstream to either use `js` instead of `js,wasm`, or `js,ecmascript` in addition to `js,wasm`.

**Comment by nevkontakte** [comment]
I'm going to give this a go, since I suspect this will have several simplification benefits for our users, as well as for maintenance. Roughly my plan is:

  - All standard library packages are built with `GOOS=js GOARCH=wasm` to maximize code reuse with upstream.
  - All user code is built with `GOOS=js GOARCH=ecmascript`.
  - Build tag `gopherjs` is always set.

At the moment I'm conducting early experiments in my own fork: https://github.com/nevkontakte/gopherjs/tree/syscall2.

**Comment by paralin** [comment]
@nevkontakte Looks good, thanks, will try to test it. following your  branch.

**Comment by nevkontakte** [comment]
Hi all, it took me longer than I wished, but I think I have a pretty usable implementation in https://github.com/nevkontakte/gopherjs/tree/syscall2 I'd like to merge into the master (before I start working on Go 1.18 support). I have a couple of minor items on the [TODO list](https://github.com/nevkontakte/gopherjs/wiki/The-Syscall-Experiment#experiment-checklist), but the hard part of the work is done.

Here are some of the notable changes:

  - The most exciting one — node-syscall extension is no longer needed for routing file system access (and a few other things like `os.Getpid()`). The old extension can still be used if you [really need](https://github.com/nevkontakte/gopherjs/blob/syscall2/doc/syscalls.md#nodejs-with-the-legacy-node-syscall-extension) it, but I plan to completely remove it in a release or two.
  - GopherJS will now standardize on `GOOS=js GOARCH=ecmascript`. This means that `GOOS` is now shared with wasm, and `GOARCH` can be used to tell them apart. In addition the new `gopherjs` build tag is always set now.
  - Standard library is built is if we were using `GOOS=js GOARCH=wasm`, meaning that we can reuse quite a bit more standard library code that works with Go WebAssembly. It should make writing portable code easier, and also reduce maintenance burden.
    - As a consequence, `syscall` package API is now the same as with `js/wasm`.
  - Output size for our reference TodoMVC app reduced by 2.17% minified, mostly thanks to throwing away unnecessary platform-specific code.
  - For folks who only run GopherJS in a browser environment, nothing significant should have changed.

While I'm polishing things up, I would like to ask everyone to try checking out my branch (https://github.com/nevkontakte/gopherjs/tree/syscall2) and trying to use that version of GopherJS with your projects. If you find any problems, please report them here.

**Comment by paralin** [comment]
@nevkontakte that is really excellent, thanks much! I'll go ahead and start using it in my projects and will let you know how it goes!

**Comment by paralin** [comment]
@nevkontakte I'm happy to report that I've switched to your syscall2 branch on a fairly large project and everything seems to work fine!

**Comment by regenvanwalbeek-wf** [comment]
@nevkontakte I tested out your syscall2 branch and was able to get everything running, though I did run into a roadbump. 

Our bindings would not load in our dev build. 

![Screen Shot 2022-03-25 at 1 09 21 PM](https://user-images.githubusercontent.com/4479792/160202400-fabf5cae-b416-4dd6-92a7-4a36d4a7db62.png)

We serve our application in the browser with [Dart Dev Compiler](https://dart.dev/tools/webdev), which generates [AMD modules](https://github.com/amdjs/amdjs-api/blob/master/AMD.md#amd) and [uses requirejs](https://github.com/amdjs/amdjs-api/blob/master/AMD.md#global-variables-) to load modules. Since [this snippet](https://github.com/nevkontakte/gopherjs/blob/efc1921fe391f2aa3a4aee6b316392accea37b8d/compiler/prelude/prelude.go#L29-L34) sees that `require` exists, it tries to load `fs` and fails. When I delete that snippet from our build output, everything loads. Do you think it's reasonable to add a try/catch, or some way to exclude that from the build? (It looks like this was taken from a previous iteration of wasm_exec.js and [that logic recently changed](https://github.com/golang/go/commit/680caf15355057ca84857a2a291b6f5c44e73329#diff-50c9536f867ce96d4731b5239b25952e38ad65800cebb9f76fbc880d34e62e94L31-L36), so I don't know how reasonable that request is)

Our production build does not use requirejs and did not run into any problems loading our js.

Other than that, looks great! We were able to remove `GOOS=linux` from our test targets (we use macs for local dev). Tests no longer error with `fatal error: all goroutines are asleep - deadlock!` 🎉  We did not discover any change in behavior in the application.

**Comment by nevkontakte** [comment]
@regenvanwalbeek-wf thanks for the report, it's a good catch. I've pushed a fix (note that it was a force push), could you try it again?

**Comment by regenvanwalbeek-wf** [comment]
@nevkontakte This fix works for us, thank you! 🚀

**Comment by paralin** [comment]
That's probably a package that isn't designed to build on js. 

You'll need to figure out which package is importing syscall and referencing those undefined types on js. I can take a look today maybe

**Comment by nevkontakte** [comment]
@flimzy this is an interesting example.

The error (obviously) comes from the `github.com/otiai10/copy` package, which sole function is (apparently) recursively copying directories. It is required transitively:

```
$ go mod why github.com/otiai10/copy
# github.com/otiai10/copy
github.com/go-kivik/kivik/v4
github.com/go-kivik/kivik/v4.test
gitlab.com/flimzy/testy
github.com/otiai10/copy
```

Looking at the source code, `github.com/otiai10/copy` seems to use `syscall` for two purposes:

1. Get atime and mtime from `io/fs.FileInfo.Sys()`. This information is present in the [wasm version of `syscall.Stat_t`](https://pkg.go.dev/syscall?GOOS=js#Stat_t), but the fields are named differently, which is unfortunate.
2. To copy a pipe via `syscall.Mkfifo()`. This operation is not supported by Go wasm, but I can see that there is a [version](https://github.com/otiai10/copy/blob/main/copy_namedpipes_x.go) of the package for platforms that don't support it.

So I think Christian is correct here, this package was never designed to work with `js/wasm` and therefore it fails for GopherJS syscall2 branch. Personally, I think it would be best to send a PR to the `github.com/otiai10/copy` adding support for wasm and upcoming GopherJS. It should be as trivial as adjusting build tags in a couple of places. We could theoretically add some extra shims on the GopherJS side to make it compile, but that would be quite hacky and go against the intent of converging with upstream as much as possible. @flimzy WDYT?

**Comment by nevkontakte** [comment]
@flimzy do you want to wait for https://github.com/otiai10/copy/pull/75 to get merged before we merge https://github.com/gopherjs/gopherjs/pull/1111?

**Comment by flimzy** [comment]
> @flimzy do you want to wait for [otiai10/copy#75](https://github.com/otiai10/copy/pull/75) to get merged before we merge #1111?

No, that's not necessary.  I have no idea how long that will take, and I can always use a fork of that repo in the meantime.

**Comment by nevkontakte** [comment]
In that case, I'm happy to merge as soon as I have your approval :)
